### PR TITLE
object: remove fetch_sub when ref_count is already 1

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
## Pull Request Description

The last `MPIR_Object_release_ref_always()` does not need an atomic `fetch_sub()`, which is heavy even if it is not contended.  This patch adds a branch to remove this `fetch_sub()` if `ref_count` is 1. This optimization is important for very fine-grained communication.

Note that an object whose `ref_count` is set to 0 will be freed, so there is no case where this `ref_count` is updated in parallel by multiple threads if `ref_count` is 1 and someone calls `MPIR_Object_release_ref_always()`.

### The performance concern of a new branch

In most cases, this branch result is deterministic in a specific function path (e.g., it is deterministic who frees a certain object in *most* cases), so we can expect that a branch prediction works well.

### The performance of atomic-fetch-sub

It highly depends on the architecture, but **under no contention**, `fetch_add()` (or `fetch_sub()`) is 5 nanoseconds (= 10 - 15 cycles), on Intel machines, while 10 nanoseconds (= about 20 cycles) on KNL, 15 nanoseconds (= about 30 cycles or more) on some ARM machines, and 20 nanoseconds (=about 50 cycles?) on IBM POWER9. This performance overhead is nonnegligible even if we consider an additional branch. At least, I believe one branch is fine to remove this.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by raffenet** [comment]
We once had a similar optimization but ended up reverting it in a5686ec3c42f0357119cab7f21df46389c7acec8 because @roblatham00 said it broke everything on Mira.

**Comment by shintaro-iwasaki** [comment]
Thanks. Um, in principle, this optimization (I mean that PR) should work, and I hope the atomic variable support gets better so we should not encounter this issue. If it would encounter an issue, I feel object handle management is wrong.

But yes, if there's a concern about this optimization since we have encountered this in the past on BG/Q Mira, I will drop this PR. This PR is not worth risking the MPI maintenance.

@raffenet Should I drop this PR?

**Comment by raffenet** [comment]
> Thanks. Um, in principle, this optimization (I mean that PR) should work, and I hope the atomic variable support gets better so we should not encounter this issue. If it would encounter an issue, I feel object handle management is wrong.
> 
> But yes, if there's a concern about this optimization since we have encountered this in the past on BG/Q Mira, I will drop this PR. This PR is not worth risking the MPICH maintenance.
> 
> @raffenet Should I drop this PR?

No, I don't think we need to drop this PR. Just digging up some historical info for context. We never identified the underlying issue with the Mira failures, just reverted and forgot about it (until now).

**Comment by hzhou** [comment]
> > @raffenet Should I drop this PR?
> 
> No, I don't think we need to drop this PR. Just digging up some historical info for context. We never identified the underlying issue with the Mira failures, just reverted and forgot about it (until now).

Yeah, let's study it and at least try to understand it a bit better even if we end up closing it.

**Comment by shintaro-iwasaki** [comment]
As far as I checked, I feel the issue is that the previous OpenPA's "atomic read" (used for a branch) is not an acquire-load (or any load that guarantees thread-thread synchronization semantics, but I don't know how BG/Q guarantees such at that point, though). I have no idea. Just a guess.

I welcome any suggestions.

**Comment by raffenet** [comment]
> As far as I checked, I feel the issue is that the previous OpenPA's "atomic read" (used for a branch) is not an acquire-load (or any load that guarantees thread-thread synchronization semantics, but I don't know how BG/Q guarantees such at that point, though). I have no idea. Just a guess.

So you think if the old version used `OPA_load_acquire_int` it might have worked? Seems like as good a guess as any to me. What would be interesting is if we could make the old, non-synchronizing, branch fail for some arch like ppc or arm to confirm.

**Comment by shintaro-iwasaki** [comment]
Yes this is my guess, though I am not fully sure (at that point there was no `OPA_load_acquire_int` and maybe a BG/Q version uses a sepcial assembly to realize `OPA_load_int`). It also highly depends on a compiler (e.g., OpenPA code puts `volatile` for everything , and actually the compiler does something for such a `volatile` variable and sometimes it happens to work like acquire-release load/store).

In my experience, although it should not be used for interthread synchronization, a normal load fails "more" on weak-memory systems like POWER or ARM even now compared with Intel machines, which sometimes make things work without proper atomic operations.

So checking this is very tricky and I cannot guarantee that I can reproduce this, but do you suggest me to try it using an old branch (`v3.2b1`) on Summit or somewhere?

**Comment by raffenet** [comment]
> So checking this is very tricky and I cannot guarantee that I can reproduce this, but do you suggest me to try it using an old branch (`v3.2b1`) on Summit or somewhere?

What I was thinking was add a commit to your branch to use `MPL_atomic_relaxed_load_int` and test on Summit or JLSE or elsewhere. Testing old MPICH might be good, but I cannot vouch for how easy it will be to build.

**Comment by hzhou** [comment]
> So checking this is very tricky and I cannot guarantee that I can reproduce this, but do you suggest me to try it using an old branch (v3.2b1) on Summit or somewhere?

I wouldn't suggest testing old branches. The codebase changed so much in the atomics and progress, I don't think it is meaningful to test. I think if we check the new branch (this PR) on JLSE's POWER and ARM, that should be sufficient.

**Comment by shintaro-iwasaki** [comment]
Okay. I don't know the consensus here, but then please let me test this branch on JLSE's POWER and ARM (I will use GCC since XLC's atomic support is not very good). I will report the results here.

**Comment by shintaro-iwasaki** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by shintaro-iwasaki** [comment]
I tried it on JLSE's POWER and ARM machines. GCC 8.3 is used. To stress atomic operations, I set `--enable-fast`.
```
sh autogen.sh
./configure --with-device=ch4:ucx --prefix=$(pwd)/install --enable-fast
make -j32 install
cd test/mpi
make testing
```
The coverage of tests was the same on both machines with and without this PR (though several tests especially that use `spawn` failed but not limited). It means that, at least, I could not observe any impact regarding thread safety at this point.

(Note: OFI did not work).

I'd be happy if you could check this PR again: @hzhou and @raffenet.

**Review by raffenet — APPROVED** [review]
LGTM. Could you add a reference to the previous implementation and the difference with yours in the commit message?

File: src/include/mpir_objects.h [hunk 1]
```diff
@@ -303,23 +303,43 @@ typedef MPL_atomic_int_t Handle_ref_count;
...
-        HANDLE_CHECK_REFCOUNT(objptr_,new_ref_,"decr");                 \
+#define MPIR_Object_release_ref_always(objptr_,inuse_ptr)                      \
+    do {                                                                       \
+        /* If it is 1, we will just free it without a heavy atomic operation.  \
+         * Note that any concurrent add_ref() to a handle whose count is 1 is  \
+         * illegal and we do not consider.                                     \
+         * (Rationale: if fetch_add() and fetch _sub() are called in parallel, \
+         *  fetch_sub() can happen to be executed earlier than fetch_add(),    \
+         *  which frees this object.  The caller of fetch_add() thinks this    \
+         * object is alive, causing invalid access. */                         \
```
Comment (hzhou) [inline src/include/mpir_objects.h:314]: Personally I don't think we need "Rationale" comment in the code. They can be in the commit messages. Rationale is for understanding related to background knowledge or reasoning. I think the comment of the case being "illegal" is sufficient for reading the code.

**Review by hzhou — APPROVED** [review]
LGTM. Just a minor comment that I don't think over-long documentation style comment is good.

**Comment by shintaro-iwasaki** [comment]
Thanks, @raffenet and @hzhou!

- I added a comment on the previous optimization (https://github.com/pmodels/mpich/commit/a5686ec3c42f0357119cab7f21df46389c7acec8)
- Remove the rationale comment.

**Comment by shintaro-iwasaki** [comment]
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by shintaro-iwasaki** [comment]
test:mpich/ch4/most

**Comment by shintaro-iwasaki** [comment]
Now it passes all the tests. I will merge this PR.

**Pull request merged**